src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/SelfDiagnostics/ApplicationInsightsEventListener.cs

```diff
@@ -0,0 +1,123 @@
+{
+    /// <summary>    
+    /// Implementation of <see cref="EventListener"/> that listens to events produced by ApplicationInsights SDK.
+    /// Logs data either every 10 seconds or when the batch becomes full, whichever occurs first.
+    /// Logs in batch to reduce the volume of logs in kusto    
+    /// </summary>
+    internal class ApplicationInsightsEventListener : EventListener
+    {
+        private static readonly DiagnosticListener _source = new DiagnosticListener(string.Concat(ApplicationInsightsDiagnosticConstants.DiagnosticSourcePrefix, nameof(ApplicationInsightsEventListener)));
+        private readonly EventLevel _eventLevel;
+
+        private const int LogFlushIntervalMs = 10 * 1000;
+        private const string EventSourceNamePrefix = "Microsoft-ApplicationInsights-";
+        private const string EventName = nameof(ApplicationInsightsEventListener);
+        private const int MaxLogLinesPerFlushInterval = 30;
+
+        private Timer _flushTimer;
+        private ConcurrentQueue<string> _logBuffer = new ConcurrentQueue<string>();
+        private ConcurrentQueue<EventSource> _eventSource = new ConcurrentQueue<EventSource>();
+        private static object _syncLock = new object();
+        private bool _disposed = false;
+
+        public ApplicationInsightsEventListener(EventLevel eventLevel)
+        {
+            this._eventLevel = eventLevel;
+            _flushTimer = new Timer
+            {
+                AutoReset = true,
+                Interval = LogFlushIntervalMs
+            };
+            _flushTimer.Elapsed += (sender, e) => Flush();
+            _flushTimer.Start();
+        }
+
+        protected override void OnEventSourceCreated(EventSource eventSource)
+        {
+            if (eventSource.Name.StartsWith(EventSourceNamePrefix))
+            {
+                EnableEvents(eventSource, _eventLevel, EventKeywords.All);
+                _eventSource.Enqueue(eventSource);
+            }
+            base.OnEventSourceCreated(eventSource);
+        }
+
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            if (!string.IsNullOrWhiteSpace(eventData.Message))
+            {
+                _logBuffer.Enqueue(string.Format(CultureInfo.InvariantCulture, eventData.Message, eventData.Payload.ToArray()));
+                if (_logBuffer.Count >= MaxLogLinesPerFlushInterval)
+                {
+                    Flush();
+                }
+            }
+        }
+
+        public void Flush()
+        {
+            if (_logBuffer.Count == 0)
+            {
+                return;
+            }
+            lock (_syncLock)
+            {
+                // batch up to 30 events in one log
+                StringBuilder sb = new StringBuilder();
+                // start with a new line
+                sb.AppendLine(string.Empty);
+                while (_logBuffer.TryDequeue(out string line))
```

Comment:
There's not anything stopping more items from coming into this `_logBuffer` while you're building up the string, is there? Seems like you could get more than 30 lines showing up here?

Reply:
As an example of a tweak you can make -- snapshot the `_logBuffer` and re-initialze it as a new queue. We do this in here, for example: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Diagnostics/FileWriter.cs#L57-L67. 

Still may end up with more than 30 entries as it's a race, but at least you know it won't log uncontrollably.

Reply:
yes, I found this during testing at high volume. Thought it's ok if the count is more than 30, there's no hard limit but it can potentially grow really big. 
Made the change.